pkg/compactor/block_upload.go

```diff
@@ -262,29 +285,73 @@ func (c *MultitenantCompactor) UploadBlockFile(w http.ResponseWriter, r *http.Re
 	w.WriteHeader(http.StatusOK)
 }
 
-func (c *MultitenantCompactor) completeBlockUpload(ctx context.Context, logger log.Logger, userBkt objstore.Bucket, blockID ulid.ULID, meta metadata.Meta) error {
+func (c *MultitenantCompactor) validateAndCompleteBlockUpload(logger log.Logger, userBkt objstore.Bucket, blockID ulid.ULID, meta *metadata.Meta) {
 	level.Debug(logger).Log("msg", "completing block upload", "files", len(meta.Thanos.Files))
 
-	// Upload meta file so block is considered complete
+	ch := make(chan struct{})
+	var wg sync.WaitGroup
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
```

Comment:
We use channel to stop `c.periodicValidationUpdater`, but it also reacts on context. Perhaps we could only use this context to control `periodicValidationUpdater` and `validateBlock`, cancel this context instead of calling `close(ch)`, and then use new context for additional calls (`uploadValidationWithError` and `markBlockComplete`)?

Reply:
Done